### ISSUE TITLE
Onchange function getting triggered on each and every component render

### ISSUE THREAD

**Issue by CV1928** [excalidraw/excalidraw#3014]
Hey,
I am using react excalidraw package.  I am trying to do collaboration feature for 2 people. Both the people will share the below component. Let's say person A and person B. If person A changes anything in the whiteboard, it will inturn trigger onChange function, which will send data over socket and will be recieved by person B in the useEffect where I am call updateScene API.
The only problem is I am running into infinite loop because onChange function is getting triggered by itself on each and every component render. Am I missing something?

This is what my component returns : 
```
  return (
<Excalidraw
          ref={props.innerChildRef}
          width={width}
          height={height}
          initialData={InitialData}
          onChange={onChange}
        />
)
```

This is the onChange function : 

```
const onChange = (elements, state) => {
    console.log("onChange:::::::::::::::::::::::::::::::::::: :", props.role, elements, "State : ", state);
    props.sendWhiteboardSocket(elements)
  };
  
```

This is my useEffect for updating elements upon receiving elements prop : 

  ```
useEffect(() => {
    if (props.elements && props.innerChildRef && props.innerChildRef.current) {
      let newElements = reconcileElements(props.elements)
      let sceneData = {"elements" : newElements}
      props.innerChildRef.current.updateScene(sceneData);
    }
  }, [props.elements, props.innerChildRef]);
```

**Comment by CV1928** [comment]
@ad1992 @dwelle  Could you please help me here?

**Comment by CV1928** [comment]
![Screenshot from 2021-02-12 15-44-18](https://user-images.githubusercontent.com/31572772/107755739-3ab65480-6d49-11eb-8e6b-171053126362.png)

Even if I directly use the react excalidraw package, on the page load itself it called onChange function twice, without doing anything in the whiteboard. Is it really the desired behaviour?

**Comment by ad1992** [comment]
@CV1928 yes currently `onChange` gets triggered everytime comp updates but this shouldn't happen. Will make the change so that it gets called only on `elements` or `appState` change.
As a workaround, you can cache the version you get using [getSceneVersion](https://github.com/excalidraw/excalidraw/tree/master/src/packages/excalidraw#getsceneversion) and check if the version has changed and perform the actions accordingly

**Comment by dwelle** [comment]
> Will make the change so that it gets called only on `elements` or `appState` change.

I don't think we should be doing that just yet. It'll slow down the update loop, especially when we rewrite the `getSceneVersion` to account for `appState`, and change the algo to be more resilient (not simply summing up element `version`s).

**Comment by ad1992** [comment]
@dwelle didn't get you. For host triggering `onChange` callback would make sense only when `elements/appstate` changes right ?  when do you think `onChange` will be needed to be called apart from these cases?

**Comment by dwelle** [comment]
Of course it shouldn't. What I'm pointing out that it'll impact perf of:

1. excalidraw.com (unless we do it inside the package).
2. Potentially even the host app, because the host apps may want to do their own diffing. For example, if they debounce whatever they're doing with the changes, there's no point in resolving the changes until needed. Or they may not care about some changes: in excalidraw.com, the `getVersion()` is super simple right now, but soon we'll  want to migrate to a more involved algo (resolve `appState`...). Host apps may not need that granularity.

If the change resolving was free, then I'd say let's do it. That said, my intuition may be off, so we should measure it — but we should measure the soon-to-be resolver, not what we do right now (sum up element verisons).

**Comment by ad1992** [comment]
AFAIU `onChange` should not affect the perf of the host and since we are passing `appState` and `elements` so unless any of those change the callback is not useful and gets triggered unnecessarily so we can try optimizing that. 
Even if we include appState in versioning still we would expect a change in `Appstate/elements` so it can be helpful.

**Comment by CV1928** [comment]
@ad1992 Any tentative time by when you can do this change. Caching also is a problem. I am having n instances of excalidraw on my webpage. Maintaining cached state for all of them is actually a headache. Is there anything else I can do here?

**Comment by ad1992** [comment]
Completely optimizing `onChange` would take some time as we need to figure out what all cases we need not call `onChange` callback. But we will roll out the optimization in patches very soon. Since you are maintaining multiple instances of Excal so you would be having some kind of mapping among the instances right so you can use that to maintain version cache mapping in memory ?

**Comment by CV1928** [comment]
@ad1992 @dwelle  Suppose I am saving excalidraw 'elements and appState' in Database and on page refresh, I am fetching that state to updateScene in useEffect. How do I make sure that the 'Excali instance is loaded' and saved in 'excalidrawRef' before I invoke updateScene Api of Excali? 
I mean useEffect can't listen to 'excalidrawRef' being updated, right?

```
useEffect(()=>{
    console.log(excalidrawRef.current) //null when useEffect runs due to 'value' change on page refresh
    if(excalidrawRef.current){
      let objectValue = JSON.parse(value) //parsed stringified value saved to db : {elements:[], appState:{}}
      excalidrawRef.current.updateScene(objectValue);
    }
  }, [value, excalidrawRe.current])

```

By the time excalidrawRef is updated, useEffect due to value has already been executed. And Since excalidrawRef.current is null at that time, It wouldn't update the scene. Now even if excalidrawRef.current is updated after the component returns JSX, it wouldn't again go to useEffect again. So scene couldn't be updated at all.

**Comment by dwelle** [comment]
You can do two things:

1. subscribe to `excalidrawRef.current.readyPromise(callback)`
2. use a `callbackRef` instead: it's what we're doing for excalidraw.com (with a bit of sugar), see https://github.com/excalidraw/excalidraw/blob/9dd2257932f458cc5a873138e51f0c745a79f187/src/excalidraw-app/index.tsx#L168-L171

**Comment by siliconjungle** [comment]
I'm having trouble with this as well, especially if the elements get set back to []. Is there a more robust way to expose versioning information?

**Comment by firmanjml** [comment]
I found a hacky way to only retrieve changes from the elements by using [use-debounce](https://www.npmjs.com/package/use-debounce). If you want to save the elements to the backend, this will help reduce number of changes and the request to the backend using ws/http req.



```
import { Excalidraw } from "@excalidraw/excalidraw";
import { ExcalidrawElement } from "@excalidraw/excalidraw/types/element/types";
import { useEffect, useState } from "react";
import { useDebounce } from 'use-debounce';

const Canvas = () => {
  const [elements, setElements] = useState<readonly ExcalidrawElement[]>([])
  const [value] = useDebounce(elements, 1000);

  useEffect(() => {
    console.log(value.filter(e => e.isDeleted === false));
  }, [value])

  return (
    <div className="excalidraw-customs" style={{ height: '100vh' }}>
      <Excalidraw theme="dark" onChange={(elementsList) => setElements(elementsList)} />
    </div >
  )
}

export default Canvas
```

**Comment by kiraaziz** [comment]
i do have the same problem 
i had a useEffect function that run each time onChange run but the problem is I do change a state in the useEffect so each time chnage the onChange rerun so the useEffect rerun so I get a infinty loop that make my app crash , so i fix it this way .
the useEffect run undependant on the onChange and run each 50 ms so I thing that sole the probelm 
`
    useEffect(() => {
        const intervalId = setInterval(() => {
            setAppState()
            setElements()
            setSelectedElements()
            setSelectedTypes()
        }, 50);

        return () => clearInterval(intervalId)
    }, [])
`

**Comment by luckyboykg** [comment]
Do we have any updates on this issue?

**Comment by brunorafaeI** [comment]
I’d like to share my version to try to fix this automatic loop trigger. I did several tests before arriving at this solution because the event, when triggered once, creates a loop of triggers within the function scope and not within the component, so I managed to resolve it this way : 


```ts
const ExcalidrawWrapper = dynamic(
  async () => (await import("@/modules/excalidraw/wrapper")).default,
  { ssr: false }
)

interface ExcalidrawModuleContentProps {
  schema: ProjectSchemaRelation & { content: ExcalidrawContent }
}

export default function ExcalidrawModuleContent({
  schema,
}: ExcalidrawModuleContentProps) {
  const [newContent, setNewContent] = useState<ExcalidrawContent | null>(null)
  const initialContent = newContent ?? schema.content
  
  const handleChangeContent = useDebounceCallback(
    async (
      elements: readonly ExcalidrawElement[],
      _: AppState,
      files: BinaryFiles
    ) => {
      const newSchemaContent = { elements: [...elements], files }
      const hasContentChanged = !isEqual(newSchemaContent, initialContent) // lodash/isEqual
  
      if (hasContentChanged) {
        setNewContent(newSchemaContent)
      }
    },
    DelayTime.EXTREME_LONG // equivalent to 3000 (3s)
  )
  
  useEffect(() => {
    if (newContent) {
      updateProjectSchema({
        variables: {
          schemaId: schema.id,
          content: newContent,
        },
      })
        .catch((errors) => {
          console.log(errors[0].message)
          toast.error(t("error"))
        })
        .finally(() => setNewContent(null))
    }
  }, [newContent, updateProjectSchema, t, schema])
  
  return (
    <ExcalidrawWrapper
      initialContent={initialContent}
      onChangeContent={handleChangeContent}
    />
  )
}
```

**Comment by abthahiahmed** [comment]
> i do have the same problem i had a useEffect function that run each time onChange run but the problem is I do change a state in the useEffect so each time chnage the onChange rerun so the useEffect rerun so I get a infinty loop that make my app crash , so i fix it this way . the useEffect run undependant on the onChange and run each 50 ms so I thing that sole the probelm `js useEffect(() => { const intervalId = setInterval(() => { setAppState() setElements() setSelectedElements() setSelectedTypes() }, 50);
> 
> ```
>     return () => clearInterval(intervalId)
> }, [])
> ```
> 
> `

This works for me. Thank you so much.
